### ISSUE TITLE
Follow operation resources

### ISSUE THREAD

**Issue by samedii** [guildai/guildai#234]
We often retrain models from previous checkpoints. I am not sure how common this workflow is for other but we use it all the time.

It would be very useful to be able to follow the runs back to be able to reproduce the results creating something like:
```
guild run train learning_rate=1e-2 max_epochs=20
guild run retrain model=... learning_rate=1e-3 max_epochs=40
guild run retrain model=... learning_rate=1e-4
```

I created something that uses guild info to try and do this:
```
import argparse
import subprocess
import json


def info(run_id):
    return json.loads(
        subprocess.run(
            f'guild runs info {run_id} --deps --json --all-scalars',
            shell=True,
            capture_output=True,
        )
        .stdout
    )


def follow_run(run_id):
    runs = [info(run_id)]
    
    while 'model' in runs[-1]['flags']:
        runs.append(
            info(runs[-1]['flags']['model'])
        )

    return runs


def pretty_print(runs):
    print(json.dumps({
        f'''{run['operation']} {index + 1}''': dict(
            id=run['id'],
            started=run['started'],
            stopped=run['stopped'],
            status=run['status'],
            train_loss=run['scalars']['tb#train/loss'],
            sourcecode_digest=run['sourcecode_digest'],
            flags=run['flags'],
        )
        for index, run in enumerate(reversed(runs))
    }, indent=4))


def command(runs):
    return '\n'.join(
        f'(\n{op}\n)'
        for op in [
            '\n'.join(
                f'\t{row}'
                for row in (
                    ['guild run ' + run['operation']]
                    + [f'{name}={value}' for name, value in run['flags'].items()]
                )
            )
            for index, run in enumerate(reversed(runs))
        ]
    )


if __name__ == '__main__':

    parser = argparse.ArgumentParser()
    parser.add_argument('run_id', type=str)
    args = parser.parse_args()

    runs = follow_run(args.run_id)
    pretty_print(runs)
    print(command(runs))
```

**Comment by gar1t** [comment]
I want to make sure I understand the motivation here. It looks like you want to re-run a particular sequence of operations. I'm guessing you apply some heuristic to determine if the results are "reproduced" - e.g. loss, etc. fall within an acceptable tolerance range. Is that right?

I would expect this to be supported with the `--proto` option, which lets you start a new run using an existing run as a prototype (i.e. use its op name, flags, and source code). I just tried this with a simple pipeline and it's not working as expected so I can't point you to a simple working example.

A proper DAG based dependency resolution scheme is planned for 0.8. This will effectively replace the so called "pipeline" operation format, which uses `steps` with a far more reasonable approach where upstream operations are resolved automatically based on the dependency graph. That's more of an aside. The `--proto` option should work as expected with pipelines in 0.7.

**Comment by samedii** [comment]
I am looking to find out the run settings that gave a certain result previously. There may have been updated data or modifications to the code since then and we want to train a new model

It doesn't sound like `proto` is what we're after but the DAG resolution scheme sounds nice

**Comment by gar1t** [comment]
You can use `--proto` to start a new run with the same settings as the specified "prototype". You can indicate you want to use the current code by including `--force-code`. Any additional flag settings are applied as well - e.g. if there's a new data set you want to train with you can override the prototype's linked data with the new run.
